distributed/tests/test_scheduler.py

```diff
@@ -2384,6 +2384,36 @@ async def test_idle_timeout_no_workers(c, s):
     assert s.check_idle()
 
 
+@gen_cluster(
+    client=True,
+    nthreads=[("127.0.0.1", 1)],
+    config={"distributed.scheduler.idle-timeout-no-worker": "1s"},
+)
+async def test_idle_timeout_unrunnable(c, s, a):
+    future = c.submit(inc, 1, workers="127.0.0.5:9999")
+
+    while not s.tasks:
+        await asyncio.sleep(0.01)
+
+    assert s.tasks[future.key] in s.unrunnable
+    assert not s.check_idle()
+
+    while not s.no_worker_since:
+        await asyncio.sleep(0.01)
+
+    # still not idle
+    assert not s.check_idle()
+
+    for _ in range(10):
+        # wait for idle-timeout-no-worker
+        await asyncio.sleep(0.1)
+
+    assert s.check_idle()
```

Comment:
This is quite fragile.

```suggestion
    await async_poll_for(s.check_idle, timeout=3)
```